Bug 1318326 - Fix `get_pings_properties` for keyedHistograms

acmiyaguchi opened the pull request:
[Bug 1318326 - |get_pings_properties| does not return all the available keyedHistograms](https://bugzilla.mozilla.org/show_bug.cgi?id=1318326)

This PR introduces failing tests that demonstrate the current behavior of `get_pings_properties` when retrieving keyedHistograms in pings that contain child processes. This is solved by defaulting to empty dictionaries to avoid `KeyError` exceptions.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/python_moztelemetry/150)
<!-- Reviewable:end -->

acmiyaguchi commented:
The updated PR fixes test failures due to bug 1363934. This also adds one more edge case that wasn't caught in the initial PR. 

The updated notebook that verifies the results can be found [here](https://gist.github.com/acmiyaguchi/0f222374f9f622b4fc3415987811ffb5).

Dexterp37 commented:
This looks good to me, but @fbertsch might be a more appropriate reviewer as I haven't touched this code in ages.

fbertsch commented on moztelemetry/spark.py:
```diff
@@ -182,16 +182,20 @@ def _get_ping_properties(ping, paths, only_median, with_processes,
                         else:
                             for payload in cursor.get("childPayloads", []):
-                                kh_keys.update(payload["keyedHistograms"][path[1]].keys())
+                                kh_keys.update(
```
Please replace all uses of update with `kh_keys |= someDict.viewkeys()`; it makes it more apparent this is a set and not a dict.

fbertsch commented on moztelemetry/spark.py:
```diff
@@ -182,16 +182,20 @@ def _get_ping_properties(ping, paths, only_median, with_processes,
                                            .get("content", {}) \
```
Mauro and I were discussing this and think that using a safe dict as the cursor might be a better (and more readable) solution. See [What is the best way to implement nested dictionaries in Python?](http://stackoverflow.com/questions/635483/what-is-the-best-way-to-implement-nested-dictionaries/19829714#19829714)

Then we can replace all `.get(elem)` with just `[elem]`

acmiyaguchi commented on moztelemetry/spark.py:
```diff
@@ -182,16 +182,20 @@ def _get_ping_properties(ping, paths, only_median, with_processes,
                         if gpu_kh is not None:
-                            kh_keys.update(gpu_kh[path[1]].keys())
+                            kh_keys.update(gpu_kh.get(path[1], {}).keys())
                 except:
```
I'm all for dropping this exception, it seems to be just catching KeyErrors.

acmiyaguchi commented:
@fbertsch This is ready for another look. I've implemented the safe dictionary and fixed any places where it could have changed the logic. I looked over the Histogram implementation, but I don't think the new KeyError-less behavior affects anything there.

fbertsch commented on moztelemetry/spark.py:
```diff
@@ -162,39 +180,34 @@ def _get_ping_properties(ping, paths, only_median, with_processes,
+                kh_keys = cursor["keyedHistograms"][path[1]].viewkeys()
+                if isinstance(cursor, dict):
+                    content_kh = cursor["processes"]["content"]["keyedHistograms"]
+                    if content_kh:
```
We don't need these ifs anymore, do we? The safe cursor will return an empty `dict`, where `content_kh[path[1]].viewkeys()` will return an empty set. So you could reduce this to:

`kh_keys |= cursor["processes"]["content"]["keyedHistograms"][path[1]].viewkeys()`

Same with `gpu_kh`.

acmiyaguchi replied:
That's true, I'll make comment of why this was structured this way.

fbertsch commented on moztelemetry/spark.py:
```diff
@@ -162,39 +180,34 @@ def _get_ping_properties(ping, paths, only_median, with_processes,
+                # property_name.
+                kh_keys = cursor["keyedHistograms"][path[1]].viewkeys()
+                if isinstance(cursor, dict):
+                    content_kh = cursor["processes"]["content"]["keyedHistograms"]
```
It occurs to me that this will fail if `cursor["processes"]`, `cursor["processes"]["content"]`, etc. are present but not `dict`s. Do you know if that schema is mandated upstream?

acmiyaguchi replied:
This should be mandated by the ingestion schema's. Here are the relevant blobs from the main ping schema that defines `processes` and `content` as objects.

https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/master/schemas/telemetry/main/main.4.schema.json#L556-L567
https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/master/schemas/telemetry/main/main.4.schema.json#L875-L876

acmiyaguchi commented:
@fbertsch this is ready for another look
